components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,409 @@
+constexpr char kToBlock[] = "toBlock";
+constexpr char kEarliestBlock[] = "earliest";
+constexpr char kLatestBlock[] = "latest";
+
+constexpr char kApprovalTopicHash[] = "Approval(address,address,uint256)";
+
+// Returns vector of chain_ids supported by allowance discovering.
+const std::vector<std::string>& GetChainIdsForAlowanceDiscovering() {
+  static base::NoDestructor<std::vector<std::string>> supported_chain_ids(
+      {brave_wallet::mojom::kMainnetChainId,
+       brave_wallet::mojom::kPolygonMainnetChainId,
+       brave_wallet::mojom::kAvalancheMainnetChainId,
+       brave_wallet::mojom::kCeloMainnetChainId,
+       brave_wallet::mojom::kArbitrumMainnetChainId,
+       brave_wallet::mojom::kOptimismMainnetChainId});
+  return *supported_chain_ids;
+}
+static std::string GetAllowanceMapKey(const std::string& contract_address,
+                                      const std::string& approver_addr,
+                                      const std::string& spender_address) {
+  return base::JoinString({contract_address, approver_addr, spender_address},
+                          "_");
+}
+
+}  // namespace
+
+namespace brave_wallet {
+EthAllowanceManager::EthAllowanceTask::EthAllowanceTask(const int& taskid)
+    : task_id(taskid) {}
+EthAllowanceManager::EthAllowanceTask::~EthAllowanceTask() = default;
+void EthAllowanceManager::EthAllowanceTask::SetResults(
+    const std::string& chain_id,
+    const uint256_t& max_block_number,
+    std::vector<mojom::AllowanceInfoPtr> alwns) {
+  allowances.insert_or_assign(
+      chain_id, std::make_tuple(max_block_number, std::move(alwns)));
+  MarkComplete();
+}
+void EthAllowanceManager::EthAllowanceTask::MarkComplete() {
+  is_completed = true;
+}
+
+EthAllowanceManager::EthAllowanceManager(BraveWalletService* wallet_service,
+                                         JsonRpcService* json_rpc_service,
+                                         KeyringService* keyring_service,
+                                         PrefService* prefs)
+    : wallet_service_(wallet_service),
+      json_rpc_service_(json_rpc_service),
+      keyring_service_(keyring_service),
+      prefs_(prefs) {}
+
+EthAllowanceManager::~EthAllowanceManager() = default;
+
+/**
+ * @brief Represents start of the allowance discovering operation
+ *
+ * @param callback callback to front end, to pass allowances list, error code
+ * and description
+ */
+void EthAllowanceManager::DiscoverEthAllowancesOnAllSupportedChains(
+    ResultCallback callback) {
+  if (!discover_eth_allowance_callback_.empty()) {
+    discover_eth_allowance_callback_.push_back(std::move(callback));
+    return;
+  }
+  discover_eth_allowance_callback_.push_back(std::move(callback));
+
+  const auto keyring_info =
+      keyring_service_->GetKeyringInfoSync(mojom::kDefaultKeyringId);
+  std::vector<std::string> account_addresses;
+  for (const auto& account_info : keyring_info->account_infos) {
+    account_addresses.push_back(account_info->address);
+  }
```

Comment:
Fresh account will not get full list allowances list, but only starting from current last block